### ISSUE TITLE
v0.9 - Filtering stats are wrong in weird ways (max indels)

### ISSUE THREAD

**Issue by isaacovercast** [dereneaton/ipyrad#358]
Changing nothing but values of max indels from 5-9. 2 Big problems.

1) The total filtered loci is _wrong_. The number of loci in the .loci file appears correct, but the counts in the stats file are far over-reported for all but the setting of 9 and above:
```
(ipyrad) bash-4.1$ grep "//" ../../lig-9indels/*.loci | wc -l
2038
(ipyrad) bash-4.1$ grep "//" ../../lig-8indels/*.loci | wc -l
1790
(ipyrad) bash-4.1$ grep "//" ../../lig-7indels/*.loci | wc -l
1509
(ipyrad) bash-4.1$ grep "//" ../../lig-6indels/*.loci | wc -l
1029
(ipyrad) bash-4.1$ grep "//" ../../lig-5indels/*.loci | wc -l
148
```

```
==> ../../lig-5indels/lig-denovo-c90_stats.txt <==

## The number of loci caught by each filter.
## ipyrad API location: [assembly].stats_dfs.s7_filters

                            total_filters  applied_order  retained_loci
total_prefiltered_loci                  0              0         157483
filtered_by_rm_duplicates            6284           6284         151199
filtered_by_max_indels               9234           9234         141965
filtered_by_max_SNPs                 7663            296         141669
filtered_by_max_shared_het           1595              8         141661
filtered_by_min_sample              97500          94162          47499
total_filtered_loci                122276         109984          47499

==> ../../lig-6indels/lig-denovo-c90_stats.txt <==

## The number of loci caught by each filter.
## ipyrad API location: [assembly].stats_dfs.s7_filters

                            total_filters  applied_order  retained_loci
total_prefiltered_loci                  0              0         157483
filtered_by_rm_duplicates            8547           8547         148936
filtered_by_max_indels              10711          10711         138225
filtered_by_max_SNPs                10640           1679         136546
filtered_by_max_shared_het           2243             51         136495
filtered_by_min_sample             127602         123157          13338
total_filtered_loci                159743         144145          13338

==> ../../lig-7indels/lig-denovo-c90_stats.txt <==

## The number of loci caught by each filter.
## ipyrad API location: [assembly].stats_dfs.s7_filters

                            total_filters  applied_order  retained_loci
total_prefiltered_loci                  0              0         157483
filtered_by_rm_duplicates            9016           9016         148467
filtered_by_max_indels               9355           9355         139112
filtered_by_max_SNPs                11166           3115         135997
filtered_by_max_shared_het           2384             71         135926
filtered_by_min_sample             136566         131914           4012
total_filtered_loci                168487         153471           4012

==> ../../lig-8indels/lig-denovo-c90_stats.txt <==

## The number of loci caught by each filter.
## ipyrad API location: [assembly].stats_dfs.s7_filters

                            total_filters  applied_order  retained_loci
total_prefiltered_loci                  0              0         157483
filtered_by_rm_duplicates            9016           9016         148467
filtered_by_max_indels               7895           7895         140572
filtered_by_max_SNPs                11166           4282         136290
filtered_by_max_shared_het           2384             83         136207
filtered_by_min_sample             136566         131914           4293
total_filtered_loci                167027         153190           4293

==> ../../lig-9indels/lig-denovo-c90_stats.txt <==


## The number of loci caught by each filter.
## ipyrad API location: [assembly].stats_dfs.s7_filters

                            total_filters  applied_order  retained_loci
total_prefiltered_loci                  0              0         157483
filtered_by_rm_duplicates            9071           9071         148412
filtered_by_max_indels               6573           6573         141839
filtered_by_max_SNPs                11198           5378         136461
filtered_by_max_shared_het           2393             97         136364
filtered_by_min_sample             139021         134326           2038
total_filtered_loci                168256         155445           2038
```

2) Only changing the value of max_indels changes the values of the `total filters` column for the other filters, which seems wrong.

**Comment by isaacovercast** [comment]
Maybe it's something about this dataset? It's PE and also like 500 samples. Still, it should work, and the way it's acting is creepy.

Allowing 10 indels is the first param setting where the results for 9 and 10 agree for all the filters like they should:
```
==> lig-10indels/lig-denovo-c90_stats.txt <==

## The number of loci caught by each filter.
## ipyrad API location: [assembly].stats_dfs.s7_filters

                            total_filters  applied_order  retained_loci
total_prefiltered_loci                  0              0         157483
filtered_by_rm_duplicates            9071           9071         148412
filtered_by_max_indels               5438           5438         142974
filtered_by_max_SNPs                11198           6369         136605
filtered_by_max_shared_het           2393            101         136504
filtered_by_min_sample             139021         134326           2178
total_filtered_loci                167121         155305           2178
```

**Comment by isaacovercast** [comment]
FUUUCCCKKKKKKKKK. I get it.

ipyrad.assemble.write_outputs:611
```
    ## If no loci survive filtering then don't dump the stats
    if np.fromiter(proc.lcov.values(), dtype=int).sum() > 0:
        # write process stats to a pickle file for collating later.
        out = {
            "filters": proc.filters,
            "lcov": proc.lcov,
            "scov": proc.scov,
            "var": proc.var,
            "pis": proc.pis,
            "nbases": proc.nbases
        }

        with open(proc.outpickle, 'wb') as outpickle:
            pickle.dump(out, outpickle)
```

and then line 275
```
        # load pickled dictionaries into a dict
        pickles = glob.glob(os.path.join(self.data.tmpdir, "*.p"))
        pdicts = {}
        for pkl in pickles:
            with open(pkl, 'rb') as inp:
                pdicts[pkl.rsplit("-", 1)[-1][:-2]] = pickle.load(inp)
```

Wait for it....... If a filtered chunk winds up with no loci remaining then the .p file isn't written, and the numbers of _filtered_ loci is seriously deflated, which results in the number of _retained_ loci going up. Mystifying and crazymaking bug.

Nasty!

Thx to @tim-oconnor for hanging in there and providing files for debugging.

Fixed in 79352a2..46f2708
